pkg/gui/controllers/helpers/fixup_helper.go

```diff
@@ -0,0 +1,175 @@
+	"github.com/jesseduffield/lazygit/pkg/gui/types"
+	"github.com/jesseduffield/lazygit/pkg/utils"
+	"github.com/samber/lo"
+)
+
+type FixupHelper struct {
+	c *HelperCommon
+}
+
+func NewFixupHelper(
+	c *HelperCommon,
+) *FixupHelper {
+	return &FixupHelper{
+		c: c,
+	}
+}
+
+type deletedLineInfo struct {
+	filename     string
+	startLineIdx int
+	numLines     int
+}
+
+func (self *FixupHelper) HandleFindBaseCommitForFixupPress() error {
+	diff, useIndex, err := self.getDiff()
+	if err != nil {
+		return err
+	}
+	if diff == "" {
+		return self.c.ErrorMsg("No changes to commit")
+	}
+
+	deletedLineInfos := self.parseDiff(diff)
+	if len(deletedLineInfos) == 0 {
+		return self.c.ErrorMsg("No deleted lines in diff")
+	}
+
+	shas := self.blameDeletedLines(deletedLineInfos)
+
+	if len(shas) == 0 {
+		// This should never happen
+		return self.c.ErrorMsg("No base commits found")
+	}
+	if len(shas) > 1 {
+		subjects, err := self.c.Git().Commit.GetShasAndCommitMessagesFirstLine(shas)
+		if err != nil {
+			return err
+		}
+		return self.c.ErrorMsg("Multiple base commits found. (" +
+			lo.Ternary(useIndex, "Try staging fewer changes", "Try staging some of the changes") +
+			")\n\n" + subjects)
+	}
+
+	commit, index, ok := lo.FindIndexOf(self.c.Model().Commits, func(commit *models.Commit) bool {
+		return commit.Sha == shas[0]
+	})
+	if !ok {
+		if self.c.Model().Commits[len(self.c.Model().Commits)-1].Status == models.StatusMerged {
+			// If the commit is not found, it's most likely because it's already
+			// merged, and more than 300 commits away. Check if the last known
+			// commit is already merged; if so, show the "already merged" error.
+			return self.c.ErrorMsg("The base commit for this change is already on master")
+		}
+		// If we get here, the current branch must have more then 300 commits. Unlikely...
+		return self.c.ErrorMsg("Base commit is not in current view")
+	}
+	if commit.Status == models.StatusMerged {
+		return self.c.ErrorMsg("The base commit for this change is already on master")
+	}
+
+	if !useIndex {
+		if err := self.c.Git().WorkingTree.StageAll(); err != nil {
+			return err
+		}
+		_ = self.c.Refresh(types.RefreshOptions{Mode: types.SYNC, Scope: []types.RefreshableView{types.FILES}})
+	}
+
+	self.c.Contexts().LocalCommits.SetSelectedLineIdx(index)
+	return self.c.PushContext(self.c.Contexts().LocalCommits)
+}
+
+func (self *FixupHelper) getDiff() (string, bool, error) {
+	args := []string{"-U0", "--ignore-submodules=all", "HEAD", "--"}
+
+	// Try staged changes first
+	useIndex := true
+	diff, err := self.c.Git().Diff.DiffIndexCmdObj(append([]string{"--cached"}, args...)...).RunWithOutput()
+
+	if err == nil && diff == "" {
+		useIndex = false
+		// If there are no staged changes, try unstaged changes
+		diff, err = self.c.Git().Diff.DiffIndexCmdObj(args...).RunWithOutput()
+	}
+
+	return diff, useIndex, err
+}
+
+func (self *FixupHelper) parseDiff(diff string) []*deletedLineInfo {
```

Comment:
Worth looking at `pkg/commands/patch/parse.go` to see if that patch parser does what you need, to spare us from having two separate patch parsers in the codebase. But if you think this use case is bespoke enough to justify its own, I'm cool with that.

Reply:
I did look at that when I first wrote the code, but found that it doesn't fit well enough for what I need here.